File: 10-heap/00621-Task_Scheduler/00621-task_scheduler.cpp
// 621: Task Scheduler
// https://leetcode.com/problems/task-scheduler


#include <vector>
#include <iostream>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> counts;
        for (char t : tasks) counts[t]++;

        priority_queue<int> pq;
        for (pair<char, int> count : counts)
            pq.push(count.second);

        int result = 0;
        int cycle = n + 1;

        while (!pq.empty()) {
            int time = 0;
            vector<int> t;

            for (int i = 0; i < cycle; i++) {
                if (!pq.empty()) {
                    t.push_back(pq.top());
                    pq.pop();
                    time++;
                }
            }

            for (int cnt : t)
                if (--cnt)
                    pq.push(cnt);

            result += !pq.empty() ? cycle : time;
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<char> tasks = {'A','C','A','B','D','B'};
    int n = 1;

    // OUTPUT
    auto result = o.leastInterval(tasks, n);
    cout << result << endl;

    return 0;
}

File: 10-heap/00621-Task_Scheduler/00621-task-scheduler.java
// 621: Task Scheduler
// https://leetcode.com/problems/task-scheduler


import java.util.PriorityQueue;
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;


class Solution {
    // SOLUTION
    public int leastInterval(char[] tasks, int n) {
        Map<Character, Integer> counts = new HashMap<Character, Integer>();
        for (char t : tasks) counts.put(t, counts.getOrDefault(t, 0) + 1);

        PriorityQueue<Integer> pq = new PriorityQueue<Integer>((a, b) -> b - a);
        pq.addAll(counts.values());

        int result = 0;
        int cycle = n + 1;

        while (!pq.isEmpty()) {
            int time = 0;
            List<Integer> tmp = new ArrayList<Integer>();

            for (int i = 0; i < cycle; i++) {
                if (!pq.isEmpty()) {
                    tmp.add(pq.poll());
                    time++;
                }
            }

            for (int cnt : tmp)
                if (--cnt > 0)
                    pq.offer(cnt);

            result += !pq.isEmpty() ? cycle : time;
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        char[] tasks = {'A','C','A','B','D','B'};
        int n = 1;

        // OUTPUT
        var result = o.leastInterval(tasks, n);
        System.out.println(result);
    }
}